compiler/noirc_evaluator/src/ssa/opt/independent_subgraph_detection.rs

```diff
@@ -0,0 +1,383 @@
+            context.brillig_return_to_argument_map.keys().copied().collect();
+
+        // Go through each disconnected set
+        for set_index in disconnected_sets_indices.iter() {
+            let current_set = &context.value_sets[*set_index];
+
+            // Find brillig-generated values in the set
+            let intersection: Vec<ValueId> =
+                all_brillig_generated_values.intersection(current_set).cloned().collect();
+            if intersection.is_empty() {
+                // This is probably a test and the values are optimized in a weird way
+                continue;
+            }
+
+            // Go through all brillig outputs in the set
+            for brillig_output_in_set in intersection.iter() {
+                // Get the inputs that correspond to the output
+                let inputs: HashSet<ValueId> = HashSet::from_iter(
+                    context.brillig_return_to_argument_map[brillig_output_in_set].iter().copied(),
+                );
+
+                // Check if any of them are not in the set
+                let unused_inputs: Vec<ValueId> = inputs.difference(current_set).copied().collect();
+
+                // There is a value not in the set, which means that the inputs/outputs of this call have not been properly constrained
+                if !unused_inputs.is_empty() {
+                    context.warnings.push(SsaReport::Bug(InternalBug::IndependentSubgraph {
+                        call_stack: function.dfg.get_call_stack(
+                            context.brillig_return_to_instruction_id_map[brillig_output_in_set],
+                        ),
+                    }));
+                }
+            }
+        }
+        context.warnings
+    }
+}
+
+#[derive(Default)]
+struct Context {
+    visited_blocks: HashSet<BasicBlockId>,
+    block_queue: Vec<BasicBlockId>,
+    warnings: Vec<SsaReport>,
+    value_sets: Vec<HashSet<ValueId>>,
+    brillig_return_to_argument_map: HashMap<ValueId, Vec<ValueId>>,
+    brillig_return_to_instruction_id_map: HashMap<ValueId, InstructionId>,
+}
+
+impl Context {
+    /// Go through each instruction in the block and add a set of ValueIds connected through that instruction
+    fn connect_value_ids_in_block(
+        &mut self,
+        function: &Function,
+        block: BasicBlockId,
+        all_functions: &BTreeMap<FunctionId, Function>,
+    ) {
+        let instructions = function.dfg[block].instructions();
+
+        for instruction in instructions.iter() {
+            let results = function.dfg.instruction_results(*instruction);
+            // For most instructions we just connect inputs and outputs
+            match &function.dfg[*instruction] {
+                Instruction::Binary(binary) => {
+                    let mut value_ids = vec![binary.lhs, binary.rhs];
+                    value_ids.extend_from_slice(results);
+                    self.connect_values(function, &value_ids);
+                }
```

Comment:
You should be able to greatly simplify this match with `Instruction::for_each_value`. You may still need an explicit match after it to check some special cases but won't need to check every case just to connect values.

Edit - just looked at connect values and realized you need to add all of these values at once together rather than separately. In that case you could probably do something like:
```rs
let mut instruction_arguments = HashSet::new();
instruction.for_each_value(|value_id| {
    if function.dfg.get_numeric_constant(value_id).is_none()) {
        instruction_arguments.insert(value_id);
    }
});
self.value_sets.push(instruction_arguments);
```

Reply:
Thanks, that's really useful